crates/pico_macros/src/memo.rs

```diff
@@ -0,0 +1,134 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, FnArg, ItemFn, PatType, ReturnType};
+
+pub(crate) fn memo(_args: TokenStream, item: TokenStream) -> TokenStream {
+    let ItemFn {
+        sig,
+        vis,
+        block,
+        attrs,
+    } = parse_macro_input!(item as ItemFn);
+
+    let fn_name = sig.ident.clone().to_string();
+
+    let inputs = sig.inputs.iter().collect::<Vec<_>>();
+
+    if inputs.is_empty() {
+        return syn::Error::new_spanned(&sig, "Function must have at least one argument")
+            .to_compile_error()
+            .into();
+    }
+
+    let db_arg = match &inputs[0] {
+        FnArg::Typed(PatType { pat, .. }) => pat.clone(),
+        _ => {
+            return syn::Error::new_spanned(inputs[0], "Expected a typed argument")
+                .to_compile_error()
+                .into()
+        }
+    };
+
+    let other_args = inputs.iter().skip(1).map(|arg| match arg {
+        FnArg::Typed(PatType { pat, .. }) => pat,
+        _ => unreachable!(),
+    });
+
+    let argument_types = inputs.iter().skip(1).map(|arg| match arg {
+        FnArg::Typed(PatType { ty, .. }) => ty,
+        _ => unreachable!(),
+    });
+
+    let return_type = match &sig.output {
+        ReturnType::Type(_, ty) => ty.clone(),
+        ReturnType::Default => {
+            return syn::Error::new_spanned(
+                &sig.output,
+                "Memoized function must have a return type",
+            )
+            .to_compile_error()
+            .into()
+        }
+    };
+
+    let (return_expr, return_type) = if let syn::Type::Reference(type_ref) = *return_type {
```

Comment:
Some questions: what if the return type is aliased, as in `type Foo = &Bar`? Is whether it is a reference evaluated at the syntactic level? Will the resulting code will fail to compile, because `downcast_ref::<&T>` doesn't work? (TBH I have no idea!)